### PR TITLE
feat(scheduling): add ReviewHistoryToEntries and simplify Reschedule for non-advanced users

### LINKED ISSUE #53: Simplify Reschedule by leveraging HistoricalMemoryStates for graded replay
## Current State

`Reschedule` takes `[]ReviewHistory` (Rating + Review timestamp + State + Due + S + D) and replays every review through `Next()` or `handleManualRating()`. It handles both graded reviews and manual ratings (set due date, suspension, reset) in a single pass.

`HistoricalMemoryStates` takes `ReviewEntries` (Rating + DeltaT) and computes `[]MemoryState` (S/D per review) via `nextStateInner`. It is simpler and only handles graded reviews.

## Proposed Enhancement

Explore using `HistoricalMemoryStates` internally within `Reschedule` for the graded review portion of the replay. This could:
- Reduce the complexity of `Reschedule` for the common case (graded-only history)
- Provide a simpler entry point for users who only need S/D reconstruction
- Keep the full `ReviewHistory` path only when manual ratings are present

## Expected Improvement

- Simpler API surface for the majority use case
- Better separation between "compute memory state" and "reconstruct full card with scheduling"
- Lower barrier to entry for non-advanced users

## Context

Suggested by @ishiko732 during review of PR #52 (Reschedule feature).

### PR DESCRIPTION
**Pull request by cantalupo555**
## Summary

Introduces a three-tier scheduling API with `RefreshDue` as the simplest entry point for the most common use case, `ReviewHistoryToEntries` + `MemoryState` for intermediate needs, and improved `Reschedule` for advanced scenarios. Based on [feedback](https://github.com/open-spaced-repetition/go-fsrs/pull/63#discussion_r3209207410) from @ishiko732 that most users just need to refresh a card's due date after parameter updates.

## Motivation

`Reschedule()` was identified as overly complex for typical use cases — open-source projects with 50+ stars that depend on ts-fsrs don't use the reschedule method. This PR provides a graduated path:

| API | Use case | Complexity |
|-----|----------|------------|
| `RefreshDue(card, reviews, now)` | "parameters changed, recalculate due date" | 1 call |
| `ReviewHistoryToEntries` → `MemoryState` | Need S/D only, no Card reconstruction | 2 calls |
| `Reschedule(card, reviews, opts)` | Manual ratings, full ReviewLog reconstruction | Advanced |

## Changes Made

- **memory.go**: Added `RefreshDue()` — chains `ReviewHistoryToEntries` → `MemoryState` → interval computation, returns updated `Card` with new `Stability`, `Difficulty`, `Due`, `ScheduledDays`, `LastReview`. Preserves `State`, `Reps`, `Lapses`, `RemainingSteps`. No `RescheduleOptions`, no `Manual` ratings, no `ReviewLog` reconstruction.
- **memory.go**: Added `ReviewHistoryToEntries()` — converts timestamp-based review history to deltaT-based entries
- **reschedule.go**: Improved input validation with indexed error messages, fixed error propagation in `calculateManualRecord`, and updated godoc to recommend `RefreshDue` as the primary API
- **fsrs_test.go**: Expanded test coverage for Reschedule edge cases (invalid ratings, zero review times, manual rating fallbacks)
- **memory_test.go**: Added `TestRefreshDue` (6 subtests: basic refresh, preserves reps/lapses/state, error cases, consistency with `MemoryState`) and `TestReviewHistoryToEntries` with cross-validation against `Reschedule`

## Testing

- [x] `go test`: all tests pass
- [x] `go vet`: no warnings
- [x] 19 new subtests covering RefreshDue, ReviewHistoryToEntries, and Reschedule edge cases

## Breaking Changes

None — purely additive. Existing `Reschedule()` behavior is unchanged.

Closes #53

---

This is an auto-generated description by sintesitech[bot]

This PR introduces a three-tier scheduling API for go-fsrs, adding `RefreshDue` as a simple entry point for recalculating due dates after parameter changes and `ReviewHistoryToEntries` for converting timestamp-based review history to deltaT entries, while improving `Reschedule` with better validation error messages. All changes are purely additive, with no breaking changes to existing behavior.

---
<sup>Written for commit c5d6b87fc1ac468f45bef1758df9fb704b82917d. Summary will update on new commits.</sup>

### CODE REVIEW

**Comment by cantalupo555** [comment]
@ishiko732 This was based on your suggestion.
Would you like any adjustments?

File: fsrs_test.go [hunk 1]
```diff
@@ -3338,6 +3339,221 @@ func TestReschedule(t *testing.T) {
...
+		}
+	})
+
+	t.Run("manual non-New as first review has zero elapsed days", func(t *testing.T) {
+		r1 := time.Date(2024, 9, 13, 0, 0, 0, 0, time.UTC)
+		manualDue := time.Date(2024, 9, 28, 0, 0, 0, 0, time.UTC)
+		reviews := []ReviewHistory{
+			{Rating: Manual, Review: r1, State: StatePtr(Review), Stability: 10.0, Difficulty: 5.0, Due: manualDue},
+		}
+
```
Comment (ishiko732) [inline fsrs_test.go:3618]: I don’t want to make it that complicated. In fact, I haven’t seen anyone manually adjust `DS`.
The intended use of reschedule is to refresh a card’s next due time after the parameters have been updated.

Open-source projects with over 50 stars that depend on ts-fsrs don’t seem to use the reschedule method.

✅ Addressed by commit [`c5d6b87`](https://github.com/open-spaced-repetition/go-fsrs/commit/c5d6b87fc1ac468f45bef1758df9fb704b82917d)

File: memory.go [hunk 2]
```diff
@@ -70,3 +70,41 @@ func (f *FSRS) MemoryState(history ReviewEntries, startingState *MemoryState) (*
...
 }
+
+// ReviewHistoryToEntries converts a timestamp-based review history into
+// DeltaT-based [ReviewEntries] suitable for use with [FSRS.MemoryState] and
+// [FSRS.HistoricalMemoryStates]. The input reviews must be in chronological
+// order.
+//
+// Returns an error if the list is empty, any review has a zero timestamp,
+// any rating is [Manual] or outside the range [Again–Easy], or any computed
+// DeltaT is negative (indicating out-of-order timestamps).
```
Comment (ishiko732) [inline memory.go:165]: That’s correct.

✅ Addressed by commit [`c5d6b87`](https://github.com/open-spaced-repetition/go-fsrs/commit/c5d6b87fc1ac468f45bef1758df9fb704b82917d)

File: memory.go [hunk 3]
```diff
@@ -108,3 +109,86 @@ func (f *FSRS) MemoryStateFromSM2(easeFactor, interval, sm2Retention float64) (*
...
+	}
+
+	card.Stability = state.Stability
+	card.Difficulty = state.Difficulty
+	card.LastReview = now
+
+	raw := f.nextIntervalRaw(state.Stability)
+	interval := math.Max(1, math.Min(math.Round(raw), f.MaximumInterval))
+	card.ScheduledDays = uint64(interval)
+	card.Due = now.Add(daysToDuration(interval, f.MaximumInterval))
```
Comment (ishiko732) [inline memory.go:153]: card.Due should be calculated based on the card’s current LastReview (or the most recent reviewTime from its reviews), rather than the current time (now). The only exception is when `card.State == New`, where using now is appropriate.

This distinction is meaningful because it allows reviews to be rescheduled when the FSRS parameters change. Recomputing due dates relative to the existing review history enables us to move reviews earlier or later without treating them as if they were reviewed again at the current moment.

**Review by ishiko732 — CHANGES_REQUESTED** [review]

File: fsrs_test.go (hunk 1 above)
Comment (cantalupo555) [reply fsrs_test.go:3618]: @ishiko732
> Open-source projects with over 50 stars that depend on ts-fsrs don’t seem to use the reschedule method.

Ironically, the project I'm working on for go-fsrs uses Reschedule 😅

**Comment by cantalupo555** [comment]
This PR is redundant, since the simplification it targets has already shipped on main (HistoricalMemoryStates / MemoryState).

Reschedule still has real consumers, the app I maintain on go-fsrs relies on its full output (Collections, Manual, RescheduleItem), so I'd rather keep Reschedule as the single advanced API (mirroring ts-fsrs's reschedule), with HistoricalMemoryStates as the go-fsrs lightweight path and drop RefreshDue as a redundant third tier.
